src/main/java/com/tabnine/inline/InlineCompletionHandler.java

```diff
@@ -0,0 +1,264 @@
+package com.tabnine.inline;
+
+import com.intellij.codeInsight.CodeInsightActionHandler;
+import com.intellij.codeInsight.completion.CompletionData;
+import com.intellij.codeInsight.completion.PlainPrefixMatcher;
+import com.intellij.codeInsight.completion.PrefixMatcher;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.EditorModificationUtil;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiFileFactory;
+import com.intellij.util.concurrency.AppExecutorUtil;
+import com.intellij.util.messages.MessageBus;
+import com.tabnine.binary.requests.autocomplete.AutocompleteResponse;
+import com.tabnine.binary.requests.autocomplete.ResultEntry;
+import com.tabnine.general.DependencyContainer;
+import com.tabnine.intellij.completions.LimitedSecletionsChangedNotifier;
+import com.tabnine.prediction.CompletionFacade;
+import com.tabnine.prediction.TabNineCompletion;
+import com.tabnine.prediction.TabNinePrefixMatcher;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static com.tabnine.general.StaticConfig.MAX_COMPLETIONS;
+import static com.tabnine.general.Utils.endsWithADot;
+
+public class InlineCompletionHandler implements CodeInsightActionHandler {
+  private static final String INLINE_DUMMY_IDENTIFIER = "TabnineInlineDummy";
```

Comment:
@dimacodota main class for handling the inline from all 3 use cases: typing, next and prev actions